src/agoradatatools/etl/utils.py

```diff
@@ -220,3 +220,116 @@ def calculate_distribution(
+        )
+
+    # Check for missing columns
+    for dataset_name, columns in required_input.items():
+        missing_columns = [
+            col for col in columns if col not in datasets[dataset_name].columns
+        ]
+        if missing_columns:
+            raise ValueError(
+                f"Missing required columns in {dataset_name} dataset: {', '.join(missing_columns)}. "
+                f"Please ensure the {dataset_name} dataset contains all required columns: {', '.join(columns)}."
+            )
+
+
+def flatten_list(lst: List) -> List:
+    """
+    Recursively flattens a nested list into a single list of values.
+
+    Args:
+        lst (list): A list which may contain nested lists at arbitrary depth.
+
+    Returns:
+        list: A new flattened list containing all the non-list elements from the input.
+
+    Example:
+        flatten(['A', ['B', 'C'], [['D'], 'E']])
+        ['A', 'B', 'C', 'D', 'E']
+    """
+    result = []
+    for item in lst:
+        if isinstance(item, list):
+            result.extend(flatten_list(item))
+        else:
+            result.append(item)
+    return result
+
+
+def remove_duplicates_keep_order(lst: List) -> List:
+    """
+    Remove duplicate elements from a list while preserving the original order.
+
+    Parameters:
+        seq (list): The input list from which to remove duplicates.
+
+    Returns:
+        list: A new list with duplicates removed, maintaining the order of first occurrence.
+
+    Example:
+        remove_duplicates(['a', 'b', 'c', 'b'])
+        ['a', 'b', 'c']
+    """
+    seen = set()
+    result = []
+    for item in lst:
+        if item not in seen:
+            seen.add(item)
+            result.append(item)
+    return result
+
+
+def create_lookup(df: pd.DataFrame, group_by_col: str) -> Dict[str, Dict[str, Any]]:
+    """
+    Creates a nested dictionary lookup from a pandas DataFrame, grouping by a specified column.
+
+    For each unique value in the specified group-by column, constructs a dictionary of the
+    remaining columns as keys and their corresponding values. If multiple rows share the same
+    group-by value but have differing values for the same column, the conflicting values are
+    merged into a list of unique values.
+
+    Args:
+        df (pd.DataFrame): The input DataFrame containing the data.
+        group_by_col (str): The column name to group the data by.
+
+    Returns:
+        Dict[str, Dict[str, Any]]: A dictionary where each key is a unique value from the
+        group-by column, and the value is another dictionary of column-value pairs.
+    """
+
+    lookup = {}
+    for _, row in df.iterrows():
+        index = row[group_by_col]
+        if index not in lookup:
+            lookup[index] = {col: row[col] for col in df.columns if col != group_by_col}
+        else:
+            for k, v in lookup[index].items():
+                if not row[k] == v:
```

Comment:
NIT: I would just use `!=` here for simplicity
```suggestion
                if row[k] != v:
```

Reply:
Oops, I originally had this structured differently and didn't notice I ended up using `not ==`. Thank you for catching this!